src/Step/Build/RunBaler.php

```diff
@@ -0,0 +1,97 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Magento\MagentoCloud\Step\Build;
+
+use Magento\MagentoCloud\Config\Stage\BuildInterface;
+use Magento\MagentoCloud\Config\ValidatorInterface;
+use Magento\MagentoCloud\Config\Validator\Result;
+use Magento\MagentoCloud\Filesystem\Flag;
+use Magento\MagentoCloud\Shell\ShellException;
+use Magento\MagentoCloud\Shell\ShellInterface;
+use Magento\MagentoCloud\Step\StepException;
+use Magento\MagentoCloud\Step\StepInterface;
+use Psr\Log\LoggerInterface;
+
+/**
+ * Attempt to perform JS bundling using Baler.
+ */
+class RunBaler implements StepInterface
+{
+    /**
+     * @var BuildInterface
+     */
+    private $buildConfig;
+
+    /**
+     * @var Flag\Manager
+     */
+    private $flagManager;
+
+    /**
+     * @var LoggerInterface
+     */
+    private $logger;
+
+    /**
+     * @var ValidatorInterface
+     */
+    private $validator;
+
+    /**
+     * @var ShellInterface
+     */
+    private $shell;
+
+    public function __construct(
+        LoggerInterface $logger,
+        BuildInterface $buildConfig,
+        Flag\Manager $flagManager,
+        ValidatorInterface $validator,
+        ShellInterface $shell
+    ) {
+        $this->logger = $logger;
+        $this->buildConfig = $buildConfig;
+        $this->flagManager = $flagManager;
+        $this->validator = $validator;
```

Comment:
You must use factory instead of passing implementation via DI

Reply:
I'm not entirely sure how you intend for me to do that? I followed the same paradigm that `Step\Build\DeployStaticContent` did in `generate.xml`.